### PR TITLE
fix: add support for parsing plpgsql bodies

### PR DESCRIPTION
**Pull request by evadnoob**
fix: add support for parsing plpgsql bodies

### CODE REVIEW

File: database/multistmt/parse.go [hunk 1]
```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
...
 	stmts := make([][]byte, 0, 1000)
+
 	for err == nil {
 		n, err := reader.Read(buf)
 		if n > 0 {
 			for i := range buf[:n] {
-				// when first two chars are comment indicators.
-				if len(buf) > 1 &&  buf[i] == '-' && buf[i+1] == '-' {
-					discard = true
+				if !fnbody {
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go:64]: Should we also check if we are inside a string literal or any other similar delimited token?

File: database/multistmt/parse.go [hunk 2]
```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
...
+					case len(buf) > 1 && i+1 < len(buf) && buf[i] == '-' && buf[i+1] == '-':
+						discard = true
+					// ignore any lines that start with // (this also covers ///)
+					case len(buf) > 1 && i+1 < len(buf) && buf[i] == '/' && buf[i+1] == '/':
+						discard = true
+					}
 				}
+
+				// output the content, for logging
 				fmt.Printf("%c", buf[i])
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go:76]: should this be removed?

File: database/postgresconn/postgresconn.go [hunk 3]
```diff
@@ -275,7 +217,12 @@ func (p *Postgres) runStatement(statement []byte) error {
 	if strings.TrimSpace(query) == "" {
 		return nil
 	}
-	if _, err := p.conn.ExecContext(ctx, query); err != nil {
+	
+	if res, err := p.conn.ExecContext(ctx, query); err != nil {
+		fmt.Printf("%+v, err: %+v\n", res, err)
```
Comment (samnguyen-outreach) [inline database/postgresconn/postgresconn.go]: should we take a proper logger instead of writing to STDOUT?

File: database/postgresconn/postgresconn.go [hunk 4]
```diff
@@ -246,16 +183,21 @@ func (p *Postgres) Unlock() error {
...
-				return false
+		var currentSchema string
+		if err := p.conn.QueryRowContext(context.Background(), "select current_schema()").Scan(&currentSchema); err != nil {
+			return err
+		}
+		fmt.Printf("current schema %v\n", currentSchema)
+		
+		if err := multistmt.Parse(migration, multiStmtDelimiter, p.config.MultiStatementMaxSize, func(m []byte) error {
+			if err := p.runStatement(m); err != nil {
+				return errors.Wrapf(err, "%s", string(m))
```
Comment (samnguyen-outreach) [inline database/postgresconn/postgresconn.go]:
```suggestion
				return errors.Wrap(err, string(m))
```

A bit simpler

File: database/multistmt/parse.go [hunk 5]
```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
...
 				switch ch := buf[i]; ch {
+				case '$':
+					// look around is there another $?
+					// is there also and ending marker like "$$ LANGUAGE plpgsql"
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go:80]:
```suggestion
					// is there also an ending marker like "$$ LANGUAGE plpgsql"
```

File: database/multistmt/parse.go [hunk 6]
```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
...
+					if buf[i+1] == '$' {
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go]: could this trigger slice out of bounds in the case where the `i` is the last index in the buffer?

File: database/multistmt/parse.go [hunk 7]
```diff
@@ -34,34 +36,65 @@ func splitWithDelimiter(delimiter []byte) func(d []byte, atEOF bool) (int, []byt
...
+						// set fnbody false to trigger the check for the next `;`
+						if fnbody {
+							fnbody = false
+						} else {
+							fnbody = true
+						}
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go]:
```suggestion
						fnbody = !fnbody
```

Simpler, but perhaps debatable if it's more readable.

File: database/multistmt/parse.go [hunk 8]
```diff
@@ -91,7 +127,10 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, h Handler)
 
 	for i, stmt := range stmts {
 		fmt.Println(i, string(stmt))
-		h(stmt)
+		if err := h(stmt); err != nil {
+			return errors.Wrapf(err, "%s", stmt)
```
Comment (samnguyen-outreach) [inline database/multistmt/parse.go:126]:
```suggestion
			return errors.Wrap(err, string(stmt))
```

File: database/postgresconn/postgresconn.go [hunk 9]
```diff
@@ -246,16 +183,21 @@ func (p *Postgres) Unlock() error {
...
 	if p.config.MultiStatementEnabled {
-		var err error
-		if e := multistmt.Parse(migration, multiStmtDelimiter, p.config.MultiStatementMaxSize, func(m []byte) bool {
-			if err = p.runStatement(m); err != nil {
-				return false
+		var currentSchema string
+		if err := p.conn.QueryRowContext(context.Background(), "select current_schema()").Scan(&currentSchema); err != nil {
+			return err
+		}
+		fmt.Printf("current schema %v\n", currentSchema)
```
Comment (samnguyen-outreach) [inline database/postgresconn/postgresconn.go:190]: should we use a logger?

File: database/postgresconn/postgresconn.go [hunk 10]
```diff
@@ -246,16 +183,21 @@ func (p *Postgres) Unlock() error {
...
+		
+		if err := multistmt.Parse(migration, multiStmtDelimiter, p.config.MultiStatementMaxSize, func(m []byte) error {
+			if err := p.runStatement(m); err != nil {
+				return errors.Wrapf(err, "%s", string(m))
 			}
-			return true
-		}); e != nil {
-			return e
+			return nil
+		}); err != nil {
```
Comment (samnguyen-outreach) [inline database/postgresconn/postgresconn.go]: I find the combined `if (expr); condition {}` construct hard to follow when split across lines, what do you think about splitting it up?   Is the purpose to narrow the scope of the `err`?

**Comment by samnguyen-outreach** [comment]
It seems like you are writing a lexer by hand from scratch, I wonder if using a lexer package could save some time?  (Perhaps too late for that...)

File: database/multistmt/parse.go (hunk 2 above)
Comment (evadnoob) [reply database/multistmt/parse.go:76]: I thought about removing it, but it's pretty helpful to see the contents of the migration

File: database/postgresconn/postgresconn.go (hunk 3 above)
Comment (evadnoob) [reply database/postgresconn/postgresconn.go]: I held off on that because this is a third party and introducing our logger seemed not quite right...

File: database/postgresconn/postgresconn.go (hunk 10 above)
Comment (evadnoob) [reply database/postgresconn/postgresconn.go]: agree, I feel like this was pre-existing code in the library, I can change

File: database/postgresconn/postgresconn.go (hunk 9 above)
Comment (evadnoob) [reply database/postgresconn/postgresconn.go:190]: I meant to take that out, will remove

File: database/multistmt/parse.go (hunk 1 above)
Comment (evadnoob) [reply database/multistmt/parse.go:64]: I think a future improvement to the parsing could cover that. I can see teams running inserts in migrations and that breaking things, good call, but future improvement.

File: database/postgresconn/postgresconn.go [hunk 11]
```diff
@@ -246,15 +183,18 @@ func (p *Postgres) Unlock() error {
...
-		var err error
-		if e := multistmt.Parse(migration, multiStmtDelimiter, p.config.MultiStatementMaxSize, func(m []byte) bool {
-			if err = p.runStatement(m); err != nil {
-				return false
-			}
-			return true
-		}); e != nil {
-			return e
+		var currentSchema string
+		if err := p.conn.QueryRowContext(context.Background(), "select current_schema()").Scan(&currentSchema); err != nil {
```
Comment (masterkidan) [inline database/postgresconn/postgresconn.go:187]: Do you want to pass the context created here to runStatement?

File: database/multistmt/parse.go [hunk 12]
```diff
@@ -41,33 +43,53 @@ func Parse(reader io.Reader, delimiter []byte, maxMigrationSize int, h Handler)
...
-				switch {
-				// ignore all lines that start with --
-				case len(buf) > 1 && i+1 < len(buf) && buf[i] == '-' && buf[i+1] == '-':
-					discard = true
-				// ignore any lines that start with // (this also covers ///)
-				case len(buf) > 1 && i+1 < len(buf) && buf[i] == '/' && buf[i+1] == '/':
-					discard = true
+				if !fnbody {
+					// when first two chars are comment indicators.
+					switch {
```
Comment (masterkidan) [inline database/multistmt/parse.go:66]: What happens if there is a comment inside a fnbody?

**Review by samnguyen-outreach — APPROVED** [review]
LGTM. Normally I don't care too much about code coverage but this seems the type of code we want very high unit test coverage over. Lots of off-by-one land mines.

File: database/multistmt/parse.go (hunk 12 above)
Comment (evadnoob) [reply database/multistmt/parse.go:66]: comments in fnbody are preserved and passed to the function(they will be part of the source) we want to keep those in source
